File: 백준/Silver/1991. 트리 순회/트리 순회.py
import sys
input = sys.stdin.readline

N = int(input())
arr = [list(input().strip().split()) for _ in range(N)]

def preorder(spot):
    if spot in tree.keys():
        print(spot, end='')
        preorder(tree[spot][0])
        preorder(tree[spot][1])
    return

def inorder(spot):
    if spot in tree.keys():
        inorder(tree[spot][0])
        print(spot, end='')
        inorder(tree[spot][1])
    return

def postorder(spot):
    if spot in tree.keys():
        postorder(tree[spot][0])
        postorder(tree[spot][1])
        print(spot, end='')
    return

tree = {}
for i in range(N):
    for j in range(1, 3):
        if arr[i][0] not in tree.keys():
            tree[arr[i][0]] = list(arr[i][j])
        else:
            tree[arr[i][0]].append(arr[i][j])
preorder('A')
print()
inorder('A')
print()
postorder('A')

File: 백준/Silver/1991. 트리 순회/트리 순회.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin')
                .toString()
                .trim()
                .split('\n');

const N = Number(input[0]);
const tree = Array.from(Array(N), () => []);
input.slice(1).forEach((e) => {
    const [edge, node1, node2] = e.split(' ');
    tree[edge.charCodeAt()-65].push(node1);        
    tree[edge.charCodeAt()-65].push(node2);        
    
});

let inorder = '';
let preorder = '';
let postorder = '';

const dfs = (edge) => {
    const [left, right] = tree[edge.charCodeAt()-65];
    preorder += edge;
    if (left !== '.') dfs(left);
    inorder += edge;
    if (right !== '.') dfs(right)
    postorder += edge;    
};

dfs('A');

console.log([preorder, inorder, postorder].join('\n'));

